### PR TITLE
Add artifact support

### PR DESCRIPTION
**Pull request by flaviuvadan**

### CODE REVIEW

File: docs/examples/workflows/upstream/artifact_disable_archive.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+with Workflow(generate_name="artifact-disable-archive-", entrypoint="artifact-disable-archive") as w:
+    whalesay = Container(
+        name="whalesay",
+        image="docker/whalesay:latest",
+        command=["sh", "-c"],
+        args=["cowsay hello world | tee /tmp/hello_world.txt | tee /tmp/hello_world_nc.txt ; sleep 1"],
+        outputs=[
+            Artifact(name="etc", path="/etc", archive=m.ArchiveStrategy(none=m.NoneStrategy())),
+            Artifact(name="hello-txt", path="/tmp/hello_world.txt", archive=m.ArchiveStrategy(none=m.NoneStrategy())),
```
Comment (samj1912) [inline docs/examples/workflows/upstream/artifact_disable_archive.md:27]: @flaviuvadan could we also add some hera classes for these strategies here?

File: docs/examples/workflows/data.md [hunk 2]
```diff
@@ -33,11 +34,7 @@ metadata:
 spec:
   templates:
   - data:
-      source:
-        artifactPaths:
-          name: test-bucket
-          s3:
-            bucket: my-bucket
```
[the remark this answers is not in the document]
Comment (samj1912) [reply docs/examples/workflows/data.md]: Fixed.

File: src/hera/workflows/steps.py [hunk 3]
```diff
@@ -38,20 +40,37 @@ class Step(
...
+    def _build_arguments(self) -> Optional[_ModelArguments]:
+        if self.arguments is None:
+            return None
+
+        artifacts = None
+        for arg in self.arguments:
+            if isinstance(arg, _ModelArtifact):
+                artifacts = [arg] if artifacts is None else [*cast(List[_ModelArtifact], artifacts), arg]
+            elif isinstance(arg, Artifact):
+                artifacts = [arg._build_artifact()] if artifacts is None else [*artifacts, arg._build_artifact()]
```
Comment (elliotgunton) [inline src/hera/workflows/steps.py:52]: IMO we can make this a bit cleaner by intialising `artifacts` to a list:
```suggestion
        artifacts = []  # may need a type hint as List[_ModelArtifact]
        for arg in self.arguments:
            if isinstance(arg, _ModelArtifact):
                artifacts.append(arg)
            elif isinstance(arg, Artifact):
                artifacts.append(arg._build_artifact())
```

And in the assignment of `model_arguments` use `artifacts=None if not artifacts else artifacts` to make use of zero-length-list's Falsey value

[Also applicable to `parameters` below]

File: src/hera/workflows/steps.py [hunk 4]
```diff
@@ -38,20 +40,37 @@ class Step(
...
+                )
+
         model_arguments = _ModelArguments(
-            artifacts=None if len(artifacts) == 0 else artifacts,
-            parameters=None if len(parameters) == 0 else parameters,
-        )
-        arguments = (
-            None if model_arguments.artifacts is None and model_arguments.parameters is None else model_arguments
+            artifacts=None if artifacts is None else artifacts,
+            parameters=None if parameters is None else parameters,
```
Comment (elliotgunton) [inline src/hera/workflows/steps.py:67]:
```suggestion
            artifacts=None if not artifacts else artifacts,
            parameters=None if not parameters else parameters,
```

**Review by samj1912 — APPROVED** [review]

File: src/hera/workflows/parameter.py [hunk 5]
```diff
@@ -54,7 +54,7 @@ def as_input(self) -> _ModelParameter:
             value_from=self.value_from,
         )
 
-    def as_argument(self) -> Optional[_ModelParameter]:
+    def as_argument(self) -> _ModelParameter:
```
Comment (elliotgunton) [inline src/hera/workflows/parameter.py:57]: Can we rename this to `as_parameter`? A bit confusing on the surface of it as we use it when creating a `parameter` under an `arguments`.

File: src/hera/workflows/workflow.py [hunk 6]
```diff
@@ -132,6 +136,34 @@ def _set_kind(cls, v):
...
+                    [arg.as_argument()]
+                    if parameters is None
+                    else [*cast(List[_ModelParameter], parameters), arg.as_argument()]
+                )
+
+        model_arguments = _ModelArguments(
+            artifacts=None if artifacts is None else artifacts,
+            parameters=None if parameters is None else parameters,
+        )
+        return None if model_arguments.artifacts is None and model_arguments.parameters is None else model_arguments
```
Comment (elliotgunton) [inline src/hera/workflows/workflow.py:165]: Cleaner code comment applicable here too

**Review by elliotgunton — COMMENTED** [review]
Awesome stuff, comments focused on the internal Hera parts for our maintenance/dev experience

File: src/hera/workflows/parameter.py (hunk 5 above)
Comment (elliotgunton) [reply src/hera/workflows/parameter.py:57]: Also as we're trying to remove the reuse of a single parameter in inputs/outputs/arguments

Comment (elliotgunton) [reply src/hera/workflows/parameter.py:57]: Deciding against this in the end to match the `as_input`/`as_output`

**Review by elliotgunton — APPROVED** [review]
🚀

File: src/hera/workflows/steps.py [hunk 7]
```diff
@@ -38,20 +40,36 @@ class Step(
...
+            return None
+
         model_arguments = _ModelArguments(
-            artifacts=None if len(artifacts) == 0 else artifacts,
-            parameters=None if len(parameters) == 0 else parameters,
-        )
-        arguments = (
-            None if model_arguments.artifacts is None and model_arguments.parameters is None else model_arguments
+            artifacts=None if not artifacts else artifacts,
+            parameters=None if not parameters else parameters,
```
Comment (elliotgunton) [inline src/hera/workflows/steps.py:66]:
```suggestion
            artifacts=artifacts or None,
            parameters=parameters or None,
```

File: src/hera/workflows/workflow.py [hunk 8]
```diff
@@ -132,6 +136,33 @@ def _set_kind(cls, v):
...
+                parameters.append(arg)
+            elif isinstance(arg, Parameter):
+                parameters.append(arg.as_argument())
+
+        if not artifacts and not parameters:
+            return None
+
+        model_arguments = _ModelArguments(
+            artifacts=None if not artifacts else artifacts,
+            parameters=None if not parameters else parameters,
```
Comment (elliotgunton) [inline src/hera/workflows/workflow.py:162]:
```suggestion
            artifacts=artifacts or None,
            parameters=parameters or None,
```

**Review by elliotgunton — COMMENTED** [review]
pythonic style 💯

**Pull request merged**